### PR TITLE
REST API for direct_messages

### PR DESCRIPTION
**Pull request by dougnukem**
support for `/direct_messages` REST API:
- https://dev.twitter.com/rest/reference/get/direct_messages

Also included unit tests.

### CODE REVIEW

**Comment by dougnukem** [comment]
ptal @dghubble when you get a chance

**Comment by mogui** [comment]
can you merge this? It is really needed :)

File: twitter/direct_messages.go [hunk 1]
```diff
@@ -1,5 +1,10 @@
 package twitter
 
+import (
+	"github.com/dghubble/sling"
```
Comment (dghubble) [inline twitter/direct_messages.go]: newline btw std lib and 3rd party

File: twitter/direct_messages.go [hunk 2]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+// newDirectMessageService returns a new StatusService.
+func newDirectMessageService(sling *sling.Sling) *DirectMessageService {
+	return &DirectMessageService{
+		baseSling: sling.New(),
+		sling:     sling.Path("direct_messages/"),
+	}
+}
+
+// DirectMessageShowParams are the parameters for DirectMessageService.Show
+type DirectMessageShowParams struct {
```
Comment (dghubble) [inline twitter/direct_messages.go]: Users won't use this with the current service structure, internal

File: twitter/direct_messages.go [hunk 3]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+	ID int64 `url:"id,omitempty"`
+}
+
+// Show returns the requested DirectMessage and all response messages.
+// https://dev.twitter.com/rest/reference/get/direct_messages/show
+func (s *DirectMessageService) Show(id int64) ([]DirectMessage, *http.Response, error) {
```
Comment (dghubble) [inline twitter/direct_messages.go]: Show returns a single Direct Message. Parsing as a list doesn't work if you try it.

File: twitter/direct_messages.go [hunk 4]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+	MaxID           int64 `url:"max_id,omitempty"`
+	Count           int   `url:"count,omitempty"`
+	IncludeEntities *bool `url:"include_entities,omitempty"`
+	SkipStatus      *bool `url:"skip_status,omitempty"`
+}
+
+// Get returns the all the direct messages
+// https://dev.twitter.com/rest/reference/get/direct_messages
+func (s *DirectMessageService) Get(params *DirectMessageGetParams) ([]DirectMessage, *http.Response, error) {
+	if params == nil {
```
Comment (dghubble) [inline twitter/direct_messages.go]: This is done for some other endpoints to merge a required value into the params. There isn't a reason to do it here. Sling's QueryStruct and BodyForm handle nil arguments.

File: twitter/direct_messages.go [hunk 5]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+// New creates a new direct message
+// https://dev.twitter.com/rest/reference/get/direct_messages/new
+func (s *DirectMessageService) New(params DirectMessageNewParams) (DirectMessage, *http.Response, error) {
+	dm := new(DirectMessage)
+	apiError := new(APIError)
+	resp, err := s.sling.New().Post("new.json").BodyForm(params).Receive(dm, apiError)
+	return *dm, resp, relevantError(err, *apiError)
+}
+
+// SendToID sends a direct message by twitter user id
```
Comment (dghubble) [inline twitter/direct_messages.go]: If these `SendToID` and `SendToScreenName` convenience methods were added, they should take params so they can accommodate changes to the DM service (e.g. if group DMs get exposed). With that parameter though, they don't seem much more valuable than just calling `New` directly.

File: twitter/direct_messages.go [hunk 6]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+
+// DirectMessageDestroyParams are the parameters for DirectMessageService.Destroy
+type DirectMessageDestroyParams struct {
+	ID              int64 `url:"id,omitempty"`
+	IncludeEntities *bool `url:"include_entities,omitempty"`
+}
+
+// Destroy deletes a direct message
+// https://dev.twitter.com/rest/reference/get/direct_messages/new
+func (s *DirectMessageService) Destroy(id int64, params *DirectMessageDestroyParams) (DirectMessage, *http.Response, error) {
```
Comment (dghubble) [inline twitter/direct_messages.go]: `go-twitter` Twitter objects are returned as pointers throughout.

File: twitter/direct_messages.go [hunk 7]
```diff
@@ -14,3 +19,119 @@ type DirectMessage struct {
...
+	if params == nil {
```
Comment (dghubble) [inline twitter/direct_messages.go]: Not needed

File: twitter/direct_messages_test.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+	client := NewClient(httpClient)
+	params := &DirectMessageGetParams{SinceID: 589147592367431680, Count: 1}
+	dms, _, err := client.DirectMessages.Get(params)
+	expected := []DirectMessage{testDM}
+	assert.Nil(t, err)
+	assert.Equal(t, expected, dms)
+}
+
+func TestDirectMessageService_GetNilParams(t *testing.T) {
```
Comment (dghubble) [inline twitter/direct_messages_test.go]: I think my use of `*_NilParams` is misleading. I have these tests for cases where required arguments need to be merged into possibly nil optional params, to test that an empty params set is created for the merge. For the DM service, these test just test the behavior of Sling's `QueryStruct` and `BodyForm` methods.

File: twitter/direct_messages_test.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+	})
+
+	client := NewClient(httpClient)
+	dms, _, err := client.DirectMessages.Sent(nil)
+	expected := []DirectMessage{testDM}
+	assert.Nil(t, err)
+	assert.Equal(t, expected, dms)
+}
+
+func TestDirectMessageService_SendToID(t *testing.T) {
```
Comment (dghubble) [inline twitter/direct_messages_test.go]: Removing the SendToID and SendToScreenName test shows that the underlying endpoint `New` isn't tested, which is disguised by the way gocover super smart. I favor removing `SendToID` and `SendToScreenName` convenience methods and just testing `New`.

If you're looking for an example of where a similar decision was made, check TimelineService.UserTimeline. It similarly requires either a user id or a user screen name. In such cases, in my opinion, it is simplest to expose the parameters and leave writing convenience wrappers to users who want it.

**Comment by dghubble** [comment]
Merged your PR as c2e8937d368, with modifications in 3cec735dcb.

I've commented this PR to explain points. Send and issue or PR if you have followups.

**Comment by mogui** [comment]
thanks :) I needed the direct message support for something I'm doing if you wanna have a [look](https://github.com/mogui/twibot)
